File: JS/496-next-greater-element-i.js
// monotonic stack

/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var nextGreaterElement = function (nums1, nums2) {
  let st = [];
  let map = new Map();
  let cur;
  for (let [i, num] of nums2.entries()) {
    while (st.length && num > nums2[st.at(-1)]) {
      cur = st.pop();
      map.set(nums2[cur], num);
    }
    st.push(i);
  }
  let res = [];
  for (let num of nums1) {
    if (map.has(num)) res.push(map.get(num));
    else res.push(-1);
  }
  return res;
};

File: C++/496-next-greater-element-i.cpp
# include "header.h"

class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        vector<int> res(nums1.size(), -1);
        if (nums1.size() == 0) return res;
        stack<int> st;
        unordered_map<int, int> umap;
        for (int i = 0; i < nums1.size(); i++) {
            umap[nums1[i]] = i;
        } 
        for (int i = 0; i < nums2.size(); i++) {
            while (!st.empty() && nums2[i] > nums2[st.top()]) {
                if (umap.find(nums2[st.top()]) != umap.end()) {
                    res[umap[nums2[st.top()]]] = nums2[i];
                }
                st.pop();
            }
            st.push(i);
        }
        return res;
    }
};